modules/fedimint-ln-common/src/internal.rs

```diff
@@ -1,50 +1,40 @@
+//! # Internal Transaction State Machine
+//!
+//! This shared state machine is used by clients
+//! that want to pay other clients within the federation
+//!
+//! It's applied in two places:
+//!   - `fedimint-ln-client` for internal payments without involving the gateway
+//!   - `gateway` for receiving payments into the federation
+
 use std::sync::Arc;
 use std::time::Duration;
 
-use bitcoin_hashes::{sha256, Hash};
+use bitcoin_hashes::sha256;
 use fedimint_client::sm::{ClientSMDatabaseTransaction, OperationId, State, StateTransition};
 use fedimint_client::transaction::ClientInput;
 use fedimint_client::DynGlobalClientContext;
-use fedimint_core::api::GlobalFederationApi;
+use fedimint_core::api::{DynModuleApi, GlobalFederationApi};
 use fedimint_core::encoding::{Decodable, Encodable};
-use fedimint_core::task::sleep;
+use fedimint_core::task::{sleep, timeout};
 use fedimint_core::{Amount, OutPoint, TransactionId};
-use fedimint_ln_client::api::LnFederationApi;
-use fedimint_ln_common::contracts::incoming::IncomingContractAccount;
-use fedimint_ln_common::contracts::{DecryptedPreimage, Preimage};
-use fedimint_ln_common::{LightningInput, LightningOutputOutcome};
 use serde::{Deserialize, Serialize};
 use thiserror::Error;
 use tracing::error;
 
-use super::{GatewayClientContext, GatewayClientStateMachines};
-use crate::gatewaylnrpc::SubscribeInterceptHtlcsResponse;
-
-#[derive(Error, Debug, Serialize, Deserialize, Encodable, Decodable, Clone, Eq, PartialEq)]
-pub enum ReceiveError {
-    #[error("Violated fee policy")]
-    ViolatedFeePolicy,
-    #[error("Invalid offer")]
-    InvalidOffer,
-    #[error("Timeout")]
-    Timeout,
-    #[error("Fetch contract error")]
-    FetchContractError,
-    #[error("Incoming contract error")]
-    IncomingContractError,
-    #[error("Invalid preimage")]
-    InvalidPreimage(Box<IncomingContractAccount>),
-    #[error("Output outcome error")]
-    OutputOutcomeError,
-    #[error("Route htlc error")]
-    RouteHtlcError,
-    #[error("Incoming contract not found")]
-    IncomingContractNotFound,
-}
+use crate::api::LnFederationApi;
+use crate::contracts::incoming::{
+    IncomingContract, IncomingContractAccount, IncomingContractOffer,
+};
+use crate::contracts::{Contract, ContractId, DecryptedPreimage, IdentifiableContract, Preimage};
+use crate::{
+    ContractOutput, LightningClientContext, LightningInput, LightningOutput, LightningOutputOutcome,
+};
 
 #[cfg_attr(doc, aquamarine::aquamarine)]
-/// State machine that executes the Lightning payment on behalf of
-/// the fedimint user that requested an invoice to be paid.
+/// State machine that executes a transaction between two users
```

Comment:
This is not necessarily true. It could be for an external lightning payment when it is used for the gateway

Reply:
Well we should consider the gateway as just another (specialized) user within the fed, no?
Occured to me that this is why we can share the same SM both in `gateway-to-client`, and `client-to-client` scenarios. In the end it's a tx between two users of a federation, but for different reasons